README.md
# Meson
### Distributed ID generator inspired by BSON's ObjectID and Snowflake

> In particle physics, mesons (/ˈmiːzɒnz/ or /ˈmɛzɒnz/) are hadronic subatomic particles composed of one quark and one antiquark, bound together by strong interactions. - [Wikipedia](https://en.wikipedia.org/wiki/Meson)


[![Maven Central](https://img.shields.io/maven-central/v/com.rfksystems/meson.svg?style=flat-square)](http://mvnrepository.com/artifact/com.rfksystems/meson)

API documentation is available [here](http://www.javadoc.io/doc/com.rfksystems/meson/)

### ID composition:

```
 48 bits | 6 bytes: Unix-time in millis as 48 bit unsigned integer 
 32 bits | 4 bytes: Generator identifier (hash)
 32 bits | 4 bytes: Generator-instance incremental sequence number (32 bit unsigned integer)
______________________________________________________________________________________________
Total 112 bits, or 14 bytes, or, 28 or 30 characters long, depending on the format
```

### Implementation details

1. Timestamps are expressed as 48 bit unsigned integers with max. value of `281474976710655`. This is well beyond what would in theory be required to store timestamps. In fact, the topmost timestamp you could store in a 48 bit int is August 2nd year 10889.
2. Timestamps have millisecond precession and all the identifier is Big-Endian exclusively. This makes Meson naturally sortable by time and in sequence within the same instance. Sorting order beyond millis for ID's generated in different instances is not defined, yet given the millisecond precession, it should not a be a problem.
3. Generator identifier is a CRC32 of parts defined, and makes a collision between hosts of the same ORG pretty near improbable:
    - `/proc/1/cgroup` contents if the file is present
    - Hostname of the "127.0.0.1" address
    - Process ID (PID)
    - Every MAC address in the system
    - Every public IPv6/IPv4 address in the system
4. Linked to Jackson's DataBind by default (at least for now), so serialization/deserialization to JSON and back should be trivial.
    
See:
- https://stackoverflow.com/a/42947044/1653859
- https://github.com/moby/moby/issues/22174
- https://tuhrig.de/how-to-know-you-are-inside-a-docker-container/
- https://docs.mongodb.com/manual/reference/method/ObjectId/
 
### Requirements

Java 8+.

### Installation

Meson is available in Maven Central. Adding dependency definition to your build config should do the trick.

#### Maven

```xml
<dependency>
    <groupId>com.rfksystems</groupId>
    <artifactId>meson</artifactId>
    <version>VERSION</version>
</dependency>
```

#### Gradle

```groovy
compile group: 'com.rfksystems', name: 'meson', version: 'VERSION'
```

### Usage

Either create new instance of `com.rfksystems.meson.Meson` with appropriate or no parameters, or call one of
`com.rfksystems.meson.Meson#directTo*` methods to create a Meson identifier directly without creating
a Meson object. See [API documentation](http://www.javadoc.io/doc/com.rfksystems/meson/) for more of what methods are available.

### Performance

... is not the main goal of the project, but as it stands, Meson is on par or quicker with BSON's ObjectID
in the limited number of tests I have performed. 

On my modest Core i5 Skylake laptop, Meson is able to generate 100m ID's on single thread in under 3 seconds -
exactly on par with BSON's ObjectId.

### Sample ID's

This should give a nice example on how the hex formatted Meson ID's look like.

#### Compact HEX string (prefered format)

```
0162915be2e1900035c91a2a5d33
0162915be2e1900035c91a2a5d34
0162915be2e1900035c91a2a5d35
0162915be2e1900035c91a2a5d36
0162915be2e1900035c91a2a5d37
0162915be2e1900035c91a2a5d38
0162915be2e1900035c91a2a5d39
0162915be2e1900035c91a2a5d3a
0162915be2e1900035c91a2a5d3b
0162915be2e1900035c91a2a5d3c
```

#### Format HEX string (display ID's)

```
0162915be2da-900035c9-1a2a5d29
0162915be2e1-900035c9-1a2a5d2a
0162915be2e1-900035c9-1a2a5d2b
0162915be2e1-900035c9-1a2a5d2c
0162915be2e1-900035c9-1a2a5d2d
0162915be2e1-900035c9-1a2a5d2e
0162915be2e1-900035c9-1a2a5d2f
0162915be2e1-900035c9-1a2a5d30
0162915be2e1-900035c9-1a2a5d31
0162915be2e1-900035c9-1a2a5d32

```

### Byte values from Java's `byte[]` output

```
[1, 98, -111, 91, -30, -31, -112, 0, 53, -55, 26, 42, 93, 61]
[1, 98, -111, 91, -30, -31, -112, 0, 53, -55, 26, 42, 93, 62]
[1, 98, -111, 91, -30, -31, -112, 0, 53, -55, 26, 42, 93, 63]
[1, 98, -111, 91, -30, -31, -112, 0, 53, -55, 26, 42, 93, 64]
[1, 98, -111, 91, -30, -31, -112, 0, 53, -55, 26, 42, 93, 65]
[1, 98, -111, 91, -30, -30, -112, 0, 53, -55, 26, 42, 93, 66]
[1, 98, -111, 91, -30, -30, -112, 0, 53, -55, 26, 42, 93, 67]
[1, 98, -111, 91, -30, -30, -112, 0, 53, -55, 26, 42, 93, 68]
[1, 98, -111, 91, -30, -30, -112, 0, 53, -55, 26, 42, 93, 69]
[1, 98, -111, 91, -30, -30, -112, 0, 53, -55, 26, 42, 93, 70]
```

## License

Apache License, Version 2.0


CHANGELOG.md
### 1.0.2

1. Allow creation of Meson identity for specific time.

### 1.0.1 

1. Comment the public API;
2. Add `_BYTES` suffix to constants `GENERATOR_ID_SIZE`, `COUNTER_SIZE`, `BUFFER_SIZE` to clarify meaning of these constants;
3. Add static methods to access current counter and generator ID's.


src/test/java/com/rfksystems/meson/MesonTest.java
package com.rfksystems.meson;

import org.bson.types.ObjectId;
import org.junit.Test;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth.assertWithMessage;


public class MesonTest {
    private static final int I100M = 100000000;
    private static final int I1M = 1000000;
    private static final byte[] TEST_GENERATOR_ID = {0x1, 0x2, 0x3, 0x4};
    private static final byte[] TEST_GENERATOR_TIME = {0x0, 0x0, 0x0, 0x0, 0x0, 0x64};
    private static final byte[] TEST_GENERATOR_SEQ = {0x0, 0x0, 0x0, 0xffffffc8};
    private static final byte[] TEST_GENERATOR_BYTES = {0x0, 0x0, 0x0, 0x0, 0x0, 0x64, 0x1, 0x2, 0x3, 0x4, 0x0, 0x0, 0x0, 0xffffffc8};

    @Test
    public void test_two_new_instances_not_equal() throws Exception {
        final Meson mesonA = new Meson();
        final Meson mesonB = new Meson();

        assertThat(mesonA).isNotEqualTo(mesonB);
    }

    @Test
    public void test_static() throws Exception {
        final byte[] bytes = Meson.directToByteArray();
        assertThat(bytes.length).isEqualTo(14);

        final ByteBuffer byteBuffer = Meson.directToByteBuffer();
        assertThat(byteBuffer.array().length).isEqualTo(14);

        final String hexString = Meson.directToHexString();
        assertThat(hexString.length()).isEqualTo(28);

        final String formatString = Meson.directToFormatString();
        assertThat(formatString.length()).isEqualTo(30);
    }

    @Test
    public void test_from_values() throws Exception {
        final Meson meson = new Meson(0x64, TEST_GENERATOR_ID, 0xc8);
        assertCommon(meson);
    }

    @Test
    public void test_from_format_string() throws Exception {
        final Meson meson = new Meson("000000000064-01020304-000000c8");
        assertCommon(meson);
    }

    @Test
    public void test_from_hex_string() throws Exception {
        final Meson meson = new Meson("00000000006401020304000000c8");
        assertCommon(meson);
    }

    @Test
    public void test_from_bytes() throws Exception {
        final Meson meson = new Meson(TEST_GENERATOR_BYTES);
        assertCommon(meson);
    }

    @Test
    public void test_from_byte_buffer() throws Exception {
        final Meson meson = new Meson(ByteBuffer.wrap(TEST_GENERATOR_BYTES));
        assertCommon(meson);
    }

    @Test
    public void test_concurrent_unique() throws Exception {
        final List<String> t1Hex = new ArrayList<>();
        final List<String> t2Hex = new ArrayList<>();
        final List<String> t3Hex = new ArrayList<>();
        final List<String> t4Hex = new ArrayList<>();

        final Runnable t1 = () -> {
            for (int i = 0; i < I1M; i++) {
                t1Hex.add(new Meson().toHexString());
            }
        };

        final Runnable t2 = () -> {
            for (int i = 0; i < I1M; i++) {
                t2Hex.add(new Meson().toHexString());
            }
        };

        final Runnable t3 = () -> {
            for (int i = 0; i < I1M; i++) {
                t3Hex.add(new Meson().toHexString());
            }
        };

        final Runnable t4 = () -> {
            for (int i = 0; i < I1M; i++) {
                t4Hex.add(new Meson().toHexString());
            }
        };

        final ExecutorService executor = Executors.newFixedThreadPool(4);

        final Future<?> submit1 = executor.submit(t1);
        final Future<?> submit2 = executor.submit(t2);
        final Future<?> submit3 = executor.submit(t3);
        final Future<?> submit4 = executor.submit(t4);

        executor.shutdown();
        executor.awaitTermination(10, TimeUnit.MINUTES);

        System.out.println("Done generating");

        assertThat(submit1.isDone()).isTrue();
        assertThat(submit2.isDone()).isTrue();
        assertThat(submit3.isDone()).isTrue();
        assertThat(submit4.isDone()).isTrue();

        assertThat(submit1.isCancelled()).isFalse();
        assertThat(submit2.isCancelled()).isFalse();
        assertThat(submit3.isCancelled()).isFalse();
        assertThat(submit4.isCancelled()).isFalse();

        assertWithMessage("T1 count is as expected").that(t1Hex).hasSize(I1M);
        assertWithMessage("T2 count is as expected").that(t2Hex).hasSize(I1M);
        assertWithMessage("T3 count is as expected").that(t3Hex).hasSize(I1M);
        assertWithMessage("T4 count is as expected").that(t4Hex).hasSize(I1M);

        assertWithMessage("T1 count is as expected").that(t1Hex.stream().distinct().count()).isEqualTo(I1M);
        assertWithMessage("T2 count is as expected").that(t2Hex.stream().distinct().count()).isEqualTo(I1M);
        assertWithMessage("T3 count is as expected").that(t3Hex.stream().distinct().count()).isEqualTo(I1M);
        assertWithMessage("T4 count is as expected").that(t4Hex.stream().distinct().count()).isEqualTo(I1M);

        final long count = Stream
                .concat(
                        Stream.concat(t1Hex.stream(), t2Hex.stream()),
                        Stream.concat(t3Hex.stream(), t4Hex.stream())
                )
                .distinct()
                .count();

        assertWithMessage("T count is as expected").that(count).isEqualTo(I1M * 4);
    }

    @Test
    public void benchmark_to_object_id() throws Exception {
        final long timeToCreateObjectId = runTimed(() -> {
            for (int i = 0; i < I100M; i++) {
                ObjectId.get();
            }
        });

        final long timeToCreateMeson = runTimed(() -> {
            for (int i = 0; i < I100M; i++) {
                new Meson();
            }
        });

        System.out.printf("Time to create 100m Meson: %dms\n", timeToCreateMeson);
        System.out.printf("Time to create 100m ObjectId: %dms\n", timeToCreateObjectId);

        final double runRatio = (double) timeToCreateMeson / (double) timeToCreateObjectId;

        assertWithMessage("Ratio of runs should be less than 2x").that(runRatio).isLessThan(2.0);
    }

    @Test
    public void test_that_it_sorts_properly() throws Exception {
        final String[] expectedOrder = {
                "0162915be2da900035c91a2a5d29",
                "0162915be2e1900035c91a2a5d33",
                "0162915be2e1900035c91a2a5d34",
                "0162915be2e1900035c91a2a5d35",
                "0162915be2e1900035c91a2a5d36",
                "0162915be2e1900035c91a2a5d37",
                "0162915be2e1900035c91a2a5d38",
                "0162915be2e1900035c91a2a5d39",
                "0162915be2e1900035c91a2a5d3a",
                "0162915be2e1900035c91a2a5d3b",
                "0162915be2e1900035c91a2a5d3c",
        };

        final ArrayList<Meson> mesons = new ArrayList<>();
        mesons.add(new Meson("0162915be2e1900035c91a2a5d36"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d37"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d38"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d39"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d3a"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d33"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d34"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d35"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d3b"));
        mesons.add(new Meson("0162915be2e1900035c91a2a5d3c"));
        mesons.add(new Meson("0162915be2da900035c91a2a5d29"));

        mesons.sort(Meson::compareTo);

        for (int i = 0; i < mesons.size(); i++) {
            assertThat(mesons.get(i).toHexString()).isEqualTo(expectedOrder[i]);
        }
    }

    private void assertCommon(final Meson meson) {
        assertThat(meson.toFormatString()).isEqualTo("000000000064-01020304-000000c8");
        assertThat(meson.toHexString()).isEqualTo("00000000006401020304000000c8");
        assertThat(meson.toByteArray()).isEqualTo(TEST_GENERATOR_BYTES);
        assertThat(meson.toByteBuffer().array()).isEqualTo(TEST_GENERATOR_BYTES);

        assertThat(meson.getTime()).isEqualTo(0x64);
        assertThat(meson.getGeneratorId()).isEqualTo(TEST_GENERATOR_ID);
        assertThat(meson.getSequence()).isEqualTo(0xc8);

        assertThat(meson.getTimeHex()).isEqualTo("000000000064");
        assertThat(meson.getGeneratorIdHex()).isEqualTo("01020304");
        assertThat(meson.getSequenceHex()).isEqualTo("000000c8");

        assertThat(meson.getTimeBytes()).isEqualTo(TEST_GENERATOR_TIME);
        assertThat(meson.getSequenceBytes()).isEqualTo(TEST_GENERATOR_SEQ);
    }

    private long runTimed(final Runnable runnable) {
        final long start = System.currentTimeMillis();
        runnable.run();
        return System.currentTimeMillis() - start;
    }
}


src/test/java/com/rfksystems/meson/JacksonTest.java
package com.rfksystems.meson;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;

public class JacksonTest {
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Test
    public void maps_to_json_and_back() throws Exception {
        final Meson meson = new Meson("01629128e71ed9686a4716e2f55e");
        final String json = objectMapper.writeValueAsString(meson);
        assertThat(json).isEqualTo("\"01629128e71ed9686a4716e2f55e\"");

        final Meson mesonFromJson = objectMapper.readValue("\"01629128e71ed9686a4716e2f55e\"", Meson.class);

        assertThat(meson).isEqualTo(mesonFromJson);
        assertThat(meson.toHexString()).isEqualTo(mesonFromJson.toHexString());
    }

    @Test
    public void maps_in_pojo_and_back() throws Exception {
        final Meson meson = new Meson("01629128e71ed9686a4716e2f55e");
        final MojoPojo mojoPojo = new MojoPojo(meson);

        assertThat(mojoPojo.getMeson()).isSameAs(meson);

        final String mojoPojoString = objectMapper.writeValueAsString(mojoPojo);

        assertThat(mojoPojoString).isEqualTo("{\"meson\":\"01629128e71ed9686a4716e2f55e\"}");

        final MojoPojo mojoPojoFromJson = objectMapper.readValue(
                "{\"meson\":\"01629128e71ed9686a4716e2f55e\"}",
                MojoPojo.class
        );

        assertThat(mojoPojo.getMeson()).isEqualTo(mojoPojoFromJson.getMeson());
        assertThat(mojoPojoFromJson.getMeson().toHexString()).isEqualTo("01629128e71ed9686a4716e2f55e");
    }

    private static final class MojoPojo {
        @JsonProperty
        private final Meson meson;

        @JsonCreator
        public MojoPojo(@JsonProperty("meson") final Meson meson) {
            this.meson = meson;
        }

        public Meson getMeson() {
            return meson;
        }
    }
}


src/main/java/com/rfksystems/meson/MesonUtils.java
/*
 * Copyright 2018 RFK Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.rfksystems.meson;

class MesonUtils {
    private static final char[] HEX_DICT = new char[]{
            '0', '1', '2', '3',
            '4', '5', '6', '7',
            '8', '9', 'a', 'b',
            'c', 'd', 'e', 'f'
    };

    static String bytesToHex(byte[] bytes) {
        final char[] hexChars = new char[bytes.length * 2];

        for (int j = 0; j < bytes.length; j++) {
            final int v = bytes[j] & 0xFF;

            hexChars[j * 2] = HEX_DICT[v >>> 4];
            hexChars[j * 2 + 1] = HEX_DICT[v & 0x0F];
        }

        return new String(hexChars);
    }

    static byte[] hexToBytes(final String hexString) {
        if (0 != hexString.length() % 2) {
            throw new IllegalArgumentException("Hex string needs to be even-length: " + hexString);
        }

        final byte[] buffer = new byte[Meson.BUFFER_SIZE_BYTES];

        for (int i = 0; i < buffer.length; i++) {
            buffer[i] = (byte) Integer.parseInt(hexString.substring(i * 2, i * 2 + 2), 16);
        }

        return buffer;
    }

    static byte[] intToBytes(final int value) {
        return new byte[]{
                (byte) ((value >> 24) & 0XFFL),
                (byte) ((value >> 16) & 0XFFL),
                (byte) ((value >> 8) & 0XFFL),
                (byte) (value & 0XFFL),
        };
    }

    static byte[] uInt48ToBytes(final long value) {
        return new byte[]{
                (byte) ((value >> 40) & 0XFFL),
                (byte) ((value >> 32) & 0XFFL),
                (byte) ((value >> 24) & 0XFFL),
                (byte) ((value >> 16) & 0XFFL),
                (byte) ((value >> 8) & 0XFFL),
                (byte) (value & 0XFFL),
        };
    }

    static byte[] longToBytes(final long value) {
        return new byte[]{
                (byte) ((value >> 56) & 0XFFL),
                (byte) ((value >> 48) & 0XFFL),
                (byte) ((value >> 40) & 0XFFL),
                (byte) ((value >> 32) & 0XFFL),
                (byte) ((value >> 24) & 0XFFL),
                (byte) ((value >> 16) & 0XFFL),
                (byte) ((value >> 8) & 0XFFL),
                (byte) (value & 0XFFL),
        };
    }

    static long longFromUInt48(byte[] bytes, final int offset) {
        return ((bytes[offset] & 0xFFL) << 40
                | (bytes[offset + 1] & 0xFFL) << 32
                | (bytes[offset + 2] & 0xFFL) << 24
                | (bytes[offset + 3] & 0xFFL) << 16
                | (bytes[offset + 4] & 0xFFL) << 8
                | (bytes[offset + 5] & 0xFFL));
    }

    static int intFromBytes(byte[] bytes, final int offset) {
        return ((bytes[offset] & 0xFF) << 24
                | (bytes[offset + 1] & 0xFF) << 16
                | (bytes[offset + 2] & 0xFF) << 8
                | (bytes[offset + 3] & 0xFF));
    }
}


src/main/java/com/rfksystems/meson/Meson.java
/*
 * Copyright 2018 RFK Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.rfksystems.meson;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.rfksystems.meson.jackson.MesonDeserializer;
import com.rfksystems.meson.jackson.MesonSerializer;

import java.io.Serializable;
import java.nio.ByteBuffer;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

import static com.rfksystems.meson.MesonUtils.*;

/**
 * Meson identity container and public API.
 */
@JsonSerialize(using = MesonSerializer.class)
@JsonDeserialize(using = MesonDeserializer.class)
public class Meson implements Serializable, Comparable<Meson> {
    /**
     * Size of the time field in number of bytes.
     */
    public static final int TIME_SIZE_BYTES = 6;

    /**
     * Size of the generator id field in number of bytes.
     */
    public static final int GENERATOR_ID_SIZE_BYTES = 4;

    /**
     * Size of the incremental sequence in number of bytes.
     */
    public static final int COUNTER_SIZE_BYTES = 4;

    /**
     * Total size of Meson identity in number of bytes.
     */
    public static final int BUFFER_SIZE_BYTES = TIME_SIZE_BYTES + GENERATOR_ID_SIZE_BYTES + COUNTER_SIZE_BYTES;

    /**
     * Topmost value for incremental sequence value, after which the value will be reset to a random number.
     */
    public static final int BORDERLINE_COUNTER_VALUE = Integer.MAX_VALUE - 1000;

    /**
     * Largest possible timestamp value Meson identity supports.
     */
    public static final long MAX_TIME = 281474976710655L;

    /**
     * Smallest possible timestamp value Meson identity supports.
     */
    public static final long MIN_TIME = 0L;

    private static final long serialVersionUID = 4304780938665765028L;

    /**
     * Instance counter. For every created Meson identity using this.
     */
    private final static AtomicInteger COUNTER = new AtomicInteger(new SecureRandom().nextInt());
    private final long time;
    private final int sequence;
    private final byte[] generatorId;

    /**
     * Create a new Meson identity for current time, with current machine identity and current sequence.
     */
    public Meson() {
        this.time = System.currentTimeMillis();
        this.sequence = getSequenceAndIncrement();
        this.generatorId = MesonGeneratorIdentity.get();
    }

    /**
     * Create a new Meson identity for given time, with current machine identity and current sequence.
     */
    public Meson(final long time) {
        if (time > MAX_TIME || time < MIN_TIME) {
            throw new IllegalArgumentException(String.format(
                    "Time must be between MIN_TIME %d and MAX_TIME %d", MIN_TIME, MAX_TIME));
        }

        this.time = time;
        this.sequence = getSequenceAndIncrement();
        this.generatorId = MesonGeneratorIdentity.get();
    }

    /**
     * Create instance of Meson identity for known time, generator identity and sequence id
     *
     * @param time        UNIX timestamp in milliseconds since epoch start, with minimum value of {@link Meson#MIN_TIME} and
     *                    maximum value of {@link Meson#MAX_TIME}.
     * @param generatorId ID of the generator, exactly {@link Meson#GENERATOR_ID_SIZE_BYTES} in size.
     * @param sequence    Sequence number of this identity, signed 32 bit integer between 0 and {@link Integer#MAX_VALUE}.
     */
    public Meson(final long time, final byte[] generatorId, final int sequence) {
        this.time = time;
        this.sequence = sequence;
        this.generatorId = generatorId;
        validate();
    }

    /**
     * Create instance of Meson identity from a hex string.
     *
     * @param string String representation of Meson identity, compact or formatted.
     */
    public Meson(final String string) {
        final String hexString = -1 == string.indexOf('-') ? string : string.replaceAll("-", "");

        if (28 != hexString.length()) {
            throw new IllegalArgumentException();
        }

        final byte[] bytes = hexToBytes(hexString);

        this.time = longFromUInt48(bytes, 0);

        this.generatorId = new byte[]{
                bytes[6],
                bytes[7],
                bytes[8],
                bytes[9]
        };

        this.sequence = intFromBytes(bytes, 10);
        validate();
    }

    /**
     * Create instance of Meson from {@link ByteBuffer} with byte representation of Meson identity.
     *
     * @param byteBuffer {@link ByteBuffer} representation of Meson identity.
     */
    public Meson(final ByteBuffer byteBuffer) {
        this(byteBuffer.array());
    }

    /**
     * Create instance of Meson from byte array with byte representation of Meson identity.
     *
     * @param bytes byte array representation of Meson identity.
     */
    public Meson(final byte[] bytes) {
        if (BUFFER_SIZE_BYTES != bytes.length) {
            throw new IllegalArgumentException();
        }

        this.time = longFromUInt48(bytes, 0);

        this.generatorId = new byte[]{
                bytes[6],
                bytes[7],
                bytes[8],
                bytes[9]
        };

        this.sequence = intFromBytes(bytes, 10);
        validate();
    }

    /**
     * Create a byte array representation of Meson identity without creating {@link Meson} object.
     *
     * @return byte array representation of Meson identity.
     */
    public static byte[] directToByteArray() {
        return toByteArray(
                System.currentTimeMillis(),
                MesonGeneratorIdentity.get(),
                getSequenceAndIncrement()
        );
    }

    /**
     * Create a {@link ByteBuffer} representation of Meson identity without creating {@link Meson} object.
     *
     * @return {@link ByteBuffer} representation of Meson identity.
     */
    public static ByteBuffer directToByteBuffer() {
        return ByteBuffer.wrap(directToByteArray());
    }

    /**
     * Create a compact-format hex String representation of Meson identity without creating {@link Meson} object.
     *
     * @return Compact-format String representation of Meson identity.
     */
    public static String directToHexString() {
        return bytesToHex(directToByteArray());
    }

    /**
     * Create a pretty-format hex String representation of Meson identity without creating {@link Meson} object.
     *
     * @return Pretty-format String representation of Meson identity.
     */
    public static String directToFormatString() {
        return bytesToHex(uInt48ToBytes(System.currentTimeMillis()))
                + "-" + bytesToHex(MesonGeneratorIdentity.get())
                + "-" + bytesToHex(intToBytes(getSequenceAndIncrement()));
    }

    /**
     * Retrieve current sequence number and increment the counter to next value.
     *
     * @return current sequence number.
     */
    private static int getSequenceAndIncrement() {
        final int sequence = COUNTER.getAndIncrement();

        if (sequence > BORDERLINE_COUNTER_VALUE) {
            COUNTER.set(new SecureRandom().nextInt());
        }

        return sequence;
    }

    /**
     * Create a byte array representation of Meson identity given time, sequence and generator ID
     *
     * @param time        UNIX timestamp in milliseconds since epoch start, with minimum value of {@link Meson#MIN_TIME} and
     *                    maximum value of {@link Meson#MAX_TIME}.
     * @param generatorId ID of the generator, exactly {@link Meson#GENERATOR_ID_SIZE_BYTES} in size.
     * @param sequence    Sequence number of this identity, signed 32 bit integer between 0 and {@link Integer#MAX_VALUE}.
     * @return byte array representation of Meson identity.
     */
    private static byte[] toByteArray(
            final long time,
            final byte[] generatorId,
            final int sequence
    ) {
        final byte[] timestampUInt48 = uInt48ToBytes(time);
        final byte[] counterInt32 = intToBytes(sequence);

        final byte[] id = new byte[BUFFER_SIZE_BYTES];
        id[0] = timestampUInt48[0];
        id[1] = timestampUInt48[1];
        id[2] = timestampUInt48[2];
        id[3] = timestampUInt48[3];
        id[4] = timestampUInt48[4];
        id[5] = timestampUInt48[5];

        id[6] = generatorId[0];
        id[7] = generatorId[1];
        id[8] = generatorId[2];
        id[9] = generatorId[3];

        id[10] = counterInt32[0];
        id[11] = counterInt32[1];
        id[12] = counterInt32[2];
        id[13] = counterInt32[3];

        return id;
    }

    /**
     * Get a number - current position of internal counter for Meson sequence numbers.
     *
     * @return current value of the counter.
     */
    public static int currentCounterValue() {
        return COUNTER.get();
    }

    /**
     * Get a byte array representation of current generator identity.
     *
     * @return byte array representation current generator identity.
     */
    public static byte[] generatorIdAsBytes() {
        return Arrays.copyOf(MesonGeneratorIdentity.get(), 4);
    }

    /**
     * Get a hex string representation of current generator identity.
     *
     * @return hex string representation current generator identity.
     */
    public static String generatorIdAsHex() {
        return bytesToHex(MesonGeneratorIdentity.get());
    }

    /**
     * Create a byte array representation of this Meson identity.
     *
     * @return byte array containing bytes that represent this Meson identity.
     */
    public byte[] toByteArray() {
        return toByteArray(time, generatorId, sequence);
    }

    /**
     * Create a {@link ByteBuffer} representation of this Meson identity.
     *
     * @return {@link ByteBuffer} containing bytes that represent this Meson identity.
     */
    public ByteBuffer toByteBuffer() {
        return ByteBuffer.wrap(toByteArray());
    }

    /**
     * Create a compact-format String representation of this Meson identity.
     *
     * @return String containing bytes that represent this Meson identity.
     */
    public String toHexString() {
        return bytesToHex(toByteArray());
    }

    /**
     * Create a pretty-format String representation of this Meson identity.
     *
     * @return String containing bytes that represent this Meson identity.
     */
    public String toFormatString() {
        return getTimeHex() + "-" + getGeneratorIdHex() + "-" + getSequenceHex();
    }

    /**
     * Get a timestamp of when this Meson identity was created.
     *
     * @return UNIX timestamp in milliseconds since epoch start, with minimum value of {@link Meson#MIN_TIME} and
     * maximum value of {@link Meson#MAX_TIME}
     */
    public long getTime() {
        return time;
    }

    /**
     * Create a instance of {@link Instant} of when this Meson identity was created.
     *
     * @return Instance of {@link Instant} that corresponds Meson identity creation time.
     */
    public Instant getInstant() {
        return Instant.ofEpochMilli(time);
    }

    /**
     * Create a instance of {@link LocalDateTime} of when this Meson identity was created.
     *
     * @return Instance of {@link LocalDateTime} that corresponds Meson identity creation time.
     */
    public LocalDateTime getLocalDateTime() {
        return LocalDateTime.from(getInstant());
    }

    /**
     * Create a byte array representation of when this Meson was created.
     *
     * @return byte array that holds a timestamp - 48-bit Big-Endian number.
     */
    public byte[] getTimeBytes() {
        return uInt48ToBytes(time);
    }

    /**
     * Create a hex encoded String representation of when this Meson was created.
     *
     * @return hex encoded String that holds a timestamp.
     */
    public String getTimeHex() {
        return bytesToHex(getTimeBytes());
    }

    /**
     * Get an integer representation of sequence number of this {@link Meson} instance.
     *
     * @return sequence number value as integer
     */
    public int getSequence() {
        return sequence;
    }

    /**
     * Get an byte array representation of sequence number of this {@link Meson} instance.
     *
     * @return sequence number value as byte array
     */
    public byte[] getSequenceBytes() {
        return intToBytes(sequence);
    }

    /**
     * Get an byte array representation of sequence number of this {@link Meson} instance.
     *
     * @return sequence number value as hex string
     */
    public String getSequenceHex() {
        return bytesToHex(getSequenceBytes());
    }

    /**
     * Get an byte array representation of generator id of this {@link Meson} instance.
     *
     * @return generator id value as byte array
     */
    public byte[] getGeneratorId() {
        return generatorId;
    }

    /**
     * Get an hex string representation of generator id of this {@link Meson} instance.
     *
     * @return generator id value as hex string
     */
    public String getGeneratorIdHex() {
        return bytesToHex(generatorId);
    }

    @Override
    public int compareTo(final Meson other) {
        if (null == other) {
            throw new NullPointerException();
        }

        final byte[] selfBytes = toByteArray();
        final byte[] otherBytes = other.toByteArray();

        for (int i = 0; i < BUFFER_SIZE_BYTES; i++) {
            if (selfBytes[i] != otherBytes[i]) {
                return ((selfBytes[i] & 0xff) < (otherBytes[i] & 0xff)) ? -1 : 1;
            }
        }

        return 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(time, sequence, Arrays.hashCode(generatorId));
    }

    @Override
    public boolean equals(final Object other) {
        if (this == other) {
            return true;
        }

        if (!(other instanceof Meson)) {
            return false;
        }

        final Meson meson = (Meson) other;

        return time == meson.time &&
                sequence == meson.sequence &&
                Arrays.equals(generatorId, meson.generatorId);
    }

    @Override
    public String toString() {
        return "Meson{" + toHexString() + '}';
    }

    private void validate() {
        if (time > MAX_TIME || time < MIN_TIME) {
            throw new IllegalArgumentException(String.format(
                    "Time must be between MIN_TIME %d and MAX_TIME %d", MIN_TIME, MAX_TIME));
        }

        if (GENERATOR_ID_SIZE_BYTES != this.generatorId.length) {
            throw new IllegalArgumentException(String.format("Generator ID must be %d bytes", GENERATOR_ID_SIZE_BYTES));
        }

        if (0 > sequence) {
            throw new IllegalArgumentException("Sequence must start from positive zero");
        }
    }
}
